evennia/contrib/slots_handler/slots.py

```diff
@@ -0,0 +1,455 @@
+"""
+Slots Handler
+
+This class is designed to sit on typeclassed objects and allow for other
+objects to be attached to configurable slots on the handler-endowed object.
+
+Built by DamnedScholar (https://github.com/damnedscholar)
+"""
+
+
+from collections import OrderedDict
+import evennia
+from evennia.objects.objects import DefaultObject
+from evennia.utils.dbserialize import _SaverDict, _SaverList, _SaverSet
+
+
+class SlotsHandler(object):
+    """
+    Handler for the slots system. This handler is designed to be attached to
+    objects with a particular @lazy-property name, so it will be referred to as
+    `.slots`, though individual games can set that however they like.
+
+    Place on character or other typeclassed object:
+    ```
+    @lazy_property
+    def slots(self):
+        return SlotsHandler(self)
+    ```
+
+    The purpose of this handler is to sit on a "holder" typeclassed object and
+    manage slots for "held" objects. The first thing that should happen is
+    using `self.slots.add()` on the holder object to identify the available
+    options. `.slots.delete()` can be used to delete or pop old slots.
+
+    By default, a held object can have slots stored on it (the handler will
+    check `.db.slots`, `.ndb.slots`, and `.slots` in that order), and
+    `self.slots.attach()` will attempt to attach it on all of those slots (it
+    should fail if it can't attach on every slot). `self.slots.drop()` will
+    remove the target from any of the slots given it. Both of these methods
+    have the option to be given a custom `slots` argument, which will override
+    the slots on the held object.
+
+    Using the custom `slots` in `attach()` and `drop()` provides some
+    customization facility in that you can store *any* object in a slot, not
+    just an object that has been set up for it.
+    """
+
+    def __init__(self, obj):
+        self.obj = obj
+        self._objid = obj.id
+
+    def __defrag_nums(self, cats):
+        "Worker function to consolidate filled numbered slots."
+        if not isinstance(cats, list):
+            cats = [cats]
+
+        arrays = self.all(obj=True)
+        arrays = [a for a in arrays if a.key in cats]
+
+        for a in arrays:
```

Comment:
See Evennia's [style guide](https://github.com/evennia/evennia/blob/master/CODING_STYLE.md#a-quick-list-of-code-style-points), use variables longer than two characters (usually three is even better as a minimum length).